CMakeLists.txt

```diff
@@ -1,120 +1,131 @@
-cmake_minimum_required(VERSION 2.8.8)
-project(instr)
+cmake_minimum_required(VERSION 3.1.0)
+project(sbt-instrumentation CXX)
+
+# --------------------------------------------------
+# Build type
+# --------------------------------------------------
+if(NOT CMAKE_BUILD_TYPE)
+  message(STATUS "Build type not set. Setting default.")
+  set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE STRING "" FORCE)
+endif()
+message(STATUS "Build type: ${CMAKE_BUILD_TYPE}")
+set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "RelWithDebInfo"
+                                                     "MinSizeRel" "Release")
 
+# --------------------------------------------------
+# Compilation flags
+# --------------------------------------------------
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF)
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -pedantic -fno-rtti")
+
+# --------------------------------------------------
+# LLVM
+# --------------------------------------------------
 find_package(LLVM REQUIRED CONFIG)
-if (${LLVM_PACKAGE_VERSION} VERSION_GREATER "3.4")
-  set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${LLVM_DIR}")
-  include(LLVMConfig)
+if(${LLVM_PACKAGE_VERSION} VERSION_GREATER "3.4")
+  list(APPEND CMAKE_MODULE_PATH "${LLVM_DIR}")
 endif()
 include(AddLLVM)
 
+# TODO: Move this to the section above when we switch to C++14
 # LLVM 10 and newer require at least C++14 standard
-if (${LLVM_PACKAGE_VERSION} VERSION_GREATER "9.0")
+if(${LLVM_PACKAGE_VERSION} VERSION_GREATER "9.0")
   set(CMAKE_CXX_STANDARD 14)
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
 else()
   # otherwise we need C++11 standard
   set(CMAKE_CXX_STANDARD 11)
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
 endif()
-set(CMAKE_CXX_STANDARD_REQUIRED ON)
-set(CMAKE_CXX_EXTENSIONS OFF)
-
-# Compiler Options
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -pedantic -fno-rtti")
 
 add_definitions(${LLVM_DEFINITIONS})
 include_directories(${LLVM_INCLUDE_DIRS})
 link_directories(${LLVM_LIBRARY_DIRS})
 
-include_directories(include)
-include_directories(analyses)
-
-find_path(LLVM_SRC_PATH include/llvm/Module.h)
-find_path(LLVM_BUILD_PATH include/llvm/Intrinsics.gen)
-
-if (LLVM_SRC_PATH)
-  include_directories(${LLVM_SRC_PATH}/include)
-  SET(LLVM_INCLUDE_DIRS ${LLVM_SRC_PATH}/include)
-  message(STATUS "Got LLVM_SRC_PATH ${LLVM_SRC_PATH}")
-endif(LLVM_SRC_PATH)
-if (LLVM_BUILD_PATH)
-  include_directories(${LLVM_BUILD_PATH}/include)
-  link_directories(${LLVM_BUILD_PATH}/lib)
-  SET(LLVM_LIBRARY_DIRS ${LLVM_BUILD_PATH}/lib)
-  message(STATUS "Got LLVM_BUILD_PATH ${LLVM_BUILD_PATH}")
-endif(LLVM_BUILD_PATH)
```

Comment:
I am not sure we want to get rid of this part. Since we still support also LLVM 3.8, the build may not provide cmake configuration and setting the source and build paths manually may be handy.

Reply:
This code section is preceded by `find_package(LLVM REQUIRED CONFIG)` so CMake would not reach these lines if the LLVM build didn't come with correct CMake modules. Besides, I've looked into the `README` and it says that LLVM 3.8 is not supported:

> To compile and run sbt-instrumentation, it is necessary to have CMake (minimal version 2.8.8) and the LLVM minimum 3.9.1 together with Clang installed.

If that's indeed true, I'll refactor the main `CMakeLists.txt` even more as almost all LLVM version checks are for compatibility with versions older than 3.9. And I'm definitely going to fix CMake version in `README` as 3.1 is the actual minimum.

Reply:
> This code section is preceded by find_package(LLVM REQUIRED CONFIG) so CMake would not reach these lines if the LLVM build didn't come with correct CMake modules. 

Good point, then let's remove it.

> Besides, I've looked into the README and it says that LLVM 3.8 is not supported:

In that case, we do not need also this bit: https://github.com/staticafi/sbt-instrumentation/pull/61/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR55

Reply:
I've rebased the PR and added an additional commit that removes remainders of old LLVM versions.